crates/dbsp/src/trace/ord/file/mod.rs

```diff
@@ -5,7 +5,69 @@ pub mod key_batch;
 pub mod val_batch;
 pub mod wset_batch;
 
+use crate::storage::file::reader::Error as ReaderError;
+use feldera_storage::error::StorageError;
 pub use indexed_wset_batch::{FileIndexedWSet, FileIndexedWSetFactories};
 pub use key_batch::{FileKeyBatch, FileKeyBatchFactories};
 pub use val_batch::{FileValBatch, FileValBatchFactories};
 pub use wset_batch::{FileWSet, FileWSetFactories};
+
+trait UnwrapStorage<T> {
+    fn unwrap_storage(self) -> T;
+}
+
+impl<T> UnwrapStorage<T> for Result<T, StorageError> {
+    fn unwrap_storage(self) -> T {
+        fn inner(error: StorageError) -> ! {
+            unwrap_storage_failed(error.to_string());
+        }
+
+        match self {
+            Ok(t) => t,
+            Err(error) => inner(error),
+        }
+    }
+}
+
+impl<T> UnwrapStorage<T> for Result<T, ReaderError> {
+    fn unwrap_storage(self) -> T {
+        fn inner(error: ReaderError) -> ! {
+            unwrap_storage_failed(error.to_string());
+        }
+
+        match self {
+            Ok(t) => t,
+            Err(error) => inner(error),
+        }
+    }
+}
+
+fn unwrap_storage_failed(error: String) -> ! {
+    panic!(
+        r#"The pipeline crashed because writing to storage failed: {error}.
+
+The crash was likely caused by storage exhaustion or storage failure, which will
```

Comment:
I suspect that only write errors can be caused by exhaustion, while read errors signal a storage failure.